crates/supervisor/core/src/reorg/task.rs

```diff
@@ -0,0 +1,397 @@
+        // Find last valid source block for this chain
+        let rewind_target_source = match self.find_rewind_target().await? {
+            Some(source) => source,
+            None => {
+                // No need to re-org for this chain
+                return Ok(());
+            }
+        };
+
+        // Get the derived block at the target source block
+        let rewind_target_derived = self.db.latest_derived_block_at_source(rewind_target_source)?;
+
+        // rewind_to() method is inclusive, so we need to get the next block.
+        let rewind_to = self.db.get_block(rewind_target_derived.number + 1)?;
+
+        // Call the rewinder to handle the DB rewinding
+        self.db.rewind(&rewind_to.id()).inspect_err(|err| {
+            warn!(
+                target: "supervisor::reorg_handler",
+                chain_id = %self.chain_id,
+                %err,
+                "Failed to rewind DB to derived block"
+            );
+        })?;
+
+        Ok(())
+    }
+
+    /// Finds the rewind target for a chain during a reorg
+    async fn find_rewind_target(&self) -> Result<Option<BlockNumHash>, SupervisorError> {
+        trace!(
+            target: "supervisor::reorg_handler",
+            chain_id = %self.chain_id,
+            "Finding rewind target..."
+        );
+
+        let latest_state = self.db.latest_derivation_state()?;
+
+        // Check if the latest source block is still canonical
+        if self.is_block_canonical(latest_state.source.number, latest_state.source.hash).await? {
+            debug!(
+                target: "supervisor::reorg_handler",
+                chain_id = %self.chain_id,
+                block_number = latest_state.source.number,
+                "Latest source block is still canonical, no reorg needed"
+            );
+            return Ok(None);
+        }
+
+        // Get finalized block and derive common ancestor
+        let finalized_block = self.db.get_safety_head_ref(kona_interop::SafetyLevel::Finalized)?;
+        let mut common_ancestor = self.db.derived_to_source(finalized_block.id())?.id();
+        let mut current_source = latest_state.source.id();
+
+        while current_source.number > common_ancestor.number {
+            // If the current source block is canonical, we found the rewind target
+            if self.is_block_canonical(current_source.number, current_source.hash).await? {
+                info!(
+                    target: "supervisor::reorg_handler",
+                    chain_id = %self.chain_id,
+                    block_number = current_source.number,
+                    "Found canonical block as rewind target"
+                );
+                common_ancestor = current_source;
+                break;
+            }
+
+            // Otherwise, walk back to the previous source block
+            current_source = self.db.get_source_block(current_source.number - 1)?.id();
+        }
```

Comment:
this loop may take a while, hence could result in an observability blind spot. let's add some partial progress logs here, e.g.
```rust
if current_source.number % X == 0 {
    trace!(
        target: "supervisor::reorg_handler",
        current_block=current_source.number,
        common_ancestor,
        "Finding rewind target..."
    )
}
```